src/core/processing/qgsnativealgorithms.cpp

```diff
@@ -2618,37 +2618,69 @@ QgsRasterLayerUniqueValuesReportAlgorithm *QgsRasterLayerUniqueValuesReportAlgor
   return new QgsRasterLayerUniqueValuesReportAlgorithm();
 }
 
-QVariantMap QgsRasterLayerUniqueValuesReportAlgorithm::processAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+bool QgsRasterLayerUniqueValuesReportAlgorithm::prepareAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback * )
 {
   QgsRasterLayer *layer = parameterAsRasterLayer( parameters, QStringLiteral( "INPUT" ), context );
   int band = parameterAsInt( parameters, QStringLiteral( "BAND" ), context );
-  QString outputFile = parameterAsFileOutput( parameters, QStringLiteral( "OUTPUT_HTML_FILE" ), context );
 
+  if ( !layer )
+    return false;
 
-  QHash< double, int > uniqueValues;
-  int width = layer->width();
-  int height = layer->height();
+  mInterface.reset( layer->dataProvider()->clone() );
+  mHasNoDataValue = layer->dataProvider()->sourceHasNoDataValue( band );
+  mLayerWidth = layer->width();
+  mLayerHeight = layer->height();
+  mExtent = layer->extent();
+  mCrs = layer->crs();
+  mRasterUnitsPerPixelX = layer->rasterUnitsPerPixelX();
+  mRasterUnitsPerPixelY = layer->rasterUnitsPerPixelY();
+  mSource = layer->source();
+
+  return true;
+}
 
-  QgsRasterBlock *rasterBlock = layer->dataProvider()->block( band, layer->extent(), width, height );
-  int noDataCount = -1;
-  if ( rasterBlock->hasNoDataValue() )
-    noDataCount = 0;
+QVariantMap QgsRasterLayerUniqueValuesReportAlgorithm::processAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  int band = parameterAsInt( parameters, QStringLiteral( "BAND" ), context );
+  QString outputFile = parameterAsFileOutput( parameters, QStringLiteral( "OUTPUT_HTML_FILE" ), context );
 
-  for ( int row = 0; row < height; row++ )
+  QHash< double, uint > uniqueValues;
+  int noDataCount = 0;
+  uint layerSize = mLayerWidth * mLayerHeight;
+  int maxWidth = 4000;
+  int maxHeight = 4000;
+  int nbBlocksWidth = std::ceil( mLayerWidth / maxWidth );
+  int nbBlocksHeight = std::ceil( mLayerHeight / maxHeight );
+  int nbBlocks = nbBlocksWidth * nbBlocksHeight;
+
+  QgsRasterIterator iter( mInterface.get() );
+  iter.setMaximumTileWidth( maxWidth );
+  iter.setMaximumTileHeight( maxHeight );
+  iter.startRasterRead( band, mLayerWidth, mLayerHeight, mExtent );
+
+  int iterLeft = 0;
+  int iterTop = 0;
+  int iterCols = 0;
+  int iterRows = 0;
+  QgsRasterBlock *rasterBlock = nullptr;
+  while ( iter.readNextRasterPart( band, iterCols, iterRows, &rasterBlock, iterLeft, iterTop ) )
```

Comment:
Careful - you're leaking rasterBlock here. You'll need to delete it before reading each next part